File: Algorithms/Classic/Permutation/permute.js
let arr1 = [1, 2, 3];
let arr2 = [4, 5, 6, 7];
let arr3 = [1, 2, 2, 3];
let arr4 = [1, 2, 3, 4, 5, 6, 7, 8, 9]

function permute(arr) {
    // 交换数组下标为 i, j 的两元素
    function swap(arr, i, j) {
        let tmp = arr[i];
        arr[i] = arr[j];
        arr[j] = tmp;
    }
    // 全排列实现
    let end = arr.length;
    let res = [];
    function wrapper(arr, begin) {
        if (begin == end) {
            console.log(arr);
            res.push(arr);
        } else {
            for (let i = begin; i < end; i++) {
                if (arr[i] == arr[i + 1]) {
                    // 如果数组的元素重复了
                    continue;
                }
                swap(arr, i, begin);
                wrapper(arr, begin + 1);
                swap(arr, i, begin);
            }
        }
    }
    wrapper(arr, 0);
}

permute(arr4);

File: Algorithms/Classic/Permutation/permute.cpp
#include <iostream>
#include <vector>

using namespace std;
vector<int> arr1 = { 1, 2, 3 };
vector<int> arr2 = { 4, 5, 6, 7 };
vector<int> arr3 = { 1, 2, 2, 3 };
vector<int> arr4 = { 1, 2, 3, 4, 5, 6, 7, 8, 9};

void swap(vector<int>& arr, int a, int b) {
 int temp = arr[a];
 arr[a] = arr[b];
 arr[b] = temp; 
}

vector<vector<int>> result;
int sum = 0;
void permute(vector<int>& arr, int length, int begin) {
  int end = arr.size();
  if(begin == end) {
    result.push_back(arr);
    for(int i : arr) {
      cout << i << " ";
    }
    cout << endl;
    sum += 1;
  } else {
    for(int i = begin; i < end; ++i) {
      if(arr[i] == arr[i + 1]) {
        continue;
      }
      swap(arr, begin, i);
      permute(arr, length, begin + 1);
      swap(arr, begin, i);
    } 
  }
}

int main(void) {
  permute(arr3, 4, 0);
  cout << "total results:" << sum;
}
